Support Attachments On File System During Export + Import

gasi-signal opened the pull request:
- [x] Load attachments into memory before export.
- [x] Save attachments to correct place in user profile upon import.
- [x] Upgrade message schema upon import.
- [x] Support both encrypted imports as well as unencrypted Chrome imports.

**Tests**

- [x] Export / import small database (25 messages) with some attachments already on disk (`schemaVersion >= 3`) and some not (`schemaVersion < 3`).
- [x] Import Chrome web app import (found and fixed issue with missing `schemaVersion`).
- [x] Export / import of real-world Signal Beta data (before: `IndexedDB` ~372MB / after: `IndexedDB` ~5.7MB / `attachments.noindex` ~234MB)
 
**Test commands:**

```javascript
// Export
let key = new Uint8Array([1, 3, 4, 5, 6, 7, 8, 11, 23, 34, 1, 34, 3, 5, 45, 45, 1, 3, 4, 5, 6, 7, 8, 11, 23, 34, 1, 34, 3, 5, 45, 45]);
Signal.Backup.exportToDirectory('/Users/<user>/Desktop/Signal-import-export-test', {key});

// Import
let key = new Uint8Array([1, 3, 4, 5, 6, 7, 8, 11, 23, 34, 1, 34, 3, 5, 45, 45, 1, 3, 4, 5, 6, 7, 8, 11, 23, 34, 1, 34, 3, 5, 45, 45]);
Signal.Backup.importFromDirectory('/Users/<user>/Desktop/Signal-import-export-test', {key});

// Import from Chrome
Signal.Backup.importFromDirectory('/Users/<user>/Desktop/Signal Export 2018 Apr 3 at 11.00.54 pm');
```

gasi-signal commented on js/modules/backup.js:
```diff
@@ -453,7 +455,7 @@ function _getAnonymousAttachmentFileName(message, index) {
 
 async function readAttachment(dir, attachment, name, options) {
   options = options || {};
-  const { key, encrypted } = options;
```
Simplified this to have the following implication `Boolean(key) ==> encrypted === true`.

gasi-signal commented on js/modules/backup.js:
```diff
@@ -492,17 +495,23 @@ async function writeAttachment(attachment, options) {
   const writer = await createFileAndWriter(dir, filename);
   const stream = createOutputStream(writer);
-  stream.write(Buffer.from(encrypted));
+  stream.write(Buffer.from(ciphertext));
```
Using `ciphertext` for consistency as there was naming overlap with `encrypted`. It’s also nicely symmetric with `plaintext`.

gasi-signal commented on js/modules/backup.js:
```diff
@@ -810,11 +819,17 @@ function saveMessage(db, message) {
 
+  const { importMessage, upgradeMessageSchema } = Signal.Migrations;
+  const importAndUpgrade = async message =>
+    upgradeMessageSchema(await importMessage(message));
```
`importMessage` preserve the existing shape and simply stores disk attachments in the right place. After that, the upgrade pipeline kicks in and takes care of upgrading the messages to the latest schema version.

scottnonnenberg-signal replied:
Thanks. This might be a good comment for the code itself.

scottnonnenberg-signal replied:
In fact, something like `writeAttachmentToDiskIfNeeded` would be more explanatory name here, without the need for the comment.

gasi-signal replied:
I like it! Always prefer to change code instead of comment if possible as it doesn’t go out of date. Now I just gotta figure out how to make that name with `create*` prefix and fitting for SSDs as well 😉

gasi-signal commented on js/modules/backup.js:
```diff
@@ -1105,7 +1118,7 @@ function deleteAll(pattern) {
 }
 
-async function backupToDirectory(directory, options) {
+async function exportToDirectory(directory, options) {
```
Renamed for symmetry as discussed 😄

gasi-signal commented on js/modules/types/attachment/migrate_data_to_file_system.js:
```diff
@@ -7,15 +7,15 @@ const {
 
 // type Context :: {
-//   writeAttachmentData :: ArrayBuffer -> Promise (IO Path)
+//   writeNewAttachmentData :: ArrayBuffer -> Promise (IO Path)
```
Added `*New*` infix to distinguish saving a new attachment (create random file name / path) vs saving an existing attachment (reuse existing `attachment.path`).

gasi-signal commented:
@scottnonnenberg-signal This is ready for review! Thanks for your help walking me through existing backup code and brainstorming best approach for import of messages with previously file system stored attachments 😄

scottnonnenberg-signal commented on js/modules/backup.js:
```diff
@@ -620,7 +629,7 @@ async function exportConversation(db, conversation, options) {
         stream.write(jsonString);
 
-        if (attachments && attachments.length) {
+        if (attachments && attachments.length > 0) {
```
What, you don't like my truthiness shorthand here? What do you have against my C heritage? ;0)

gasi-signal replied:
Vintage™!

scottnonnenberg-signal commented on js/modules/types/message.js:
```diff
@@ -166,13 +166,72 @@ const toVersion3 = exports._withSchemaVersion(
+    const lastVersionWithAttachmentDataInMemory = 2;
+    const willHaveAttachmentsSavedOnFileSystemDuringUpgrade =
+      message.schemaVersion <= lastVersionWithAttachmentDataInMemory;
+    if (willHaveAttachmentsSavedOnFileSystemDuringUpgrade) {
```
Very explanatory naming. Much clarity. Many understandings.

scottnonnenberg-signal replied:
An slightly shorter name, just food for thought: `willAttachmentsGoToDiskOnUpgrade`

gasi-signal replied:
LOL! I like `willAttachmentsGoToDiskOnUpgrade` but I struggled with `*Disk*` myself wondering if SSDs wouldn’t count 😜

gasi-signal commented:
@scottnonnenberg-signal Thanks for the review! Incorporated your feedback and added a helper `start-backup` script. Rebased on top of `development` and CR feedback changes: https://github.com/signalapp/Signal-Desktop/pull/2212/files/5ea1a305a69cdf2c3a82cb9b289df547c70cb800..b680b61843a858cd016ac32053ecc986413b1ad0
